runtime/src/iree/hal/cts/testdata/command_buffer_push_constants_test.mlir

```diff
@@ -14,12 +14,12 @@ hal.executable.source public @executable {
   }
   builtin.module {
     func.func @write_push_constants() {
-      %input_0 = hal.interface.constant.load[0] : i32
-      %input_1 = hal.interface.constant.load[1] : i32
-      %input_2 = hal.interface.constant.load[2] : i32
-      %input_3 = hal.interface.constant.load[3] : i32
+      %input_0 = hal.interface.constant.load layout(#pipeline_layout) ordinal(0) : i32
+      %input_1 = hal.interface.constant.load layout(#pipeline_layout) ordinal(1) : i32
+      %input_2 = hal.interface.constant.load layout(#pipeline_layout) ordinal(2) : i32
+      %input_3 = hal.interface.constant.load layout(#pipeline_layout) ordinal(3) : i32
```

Comment:
Nice having these explicit. Needing to look them up by walking the IR to find an export op was annoying :P

Can we ever end up using multiple pipeline layouts at this point? Maybe we could have a verifier that enforces only one?

```mlir
#pipeline_layout_0 = #hal.pipeline.layout<push_constants = 4, sets = [
  #hal.descriptor_set.layout<0, bindings = [
    #hal.descriptor_set.binding<0, storage_buffer>
  ]>
]>

#pipeline_layout_1 = #hal.pipeline.layout<push_constants = 2, sets = [
  #hal.descriptor_set.layout<0, bindings = [
    #hal.descriptor_set.binding<0, storage_buffer>
  ]>
]>

hal.executable.source public @executable {
  hal.executable.export public @write_push_constants ordinal(0) layout(#pipeline_layout) attributes {workgroup_size = [1 : index, 1 : index, 1 : index]} {
  ^bb0(%arg0: !hal.device):
    %c1 = arith.constant 1 : index
    hal.return %c1, %c1, %c1 : index, index, index
  }
  builtin.module {
    func.func @write_push_constants() {
      %input_0 = hal.interface.constant.load layout(#pipeline_layout_0) ordinal(0) : i32
      %input_1 = hal.interface.constant.load layout(#pipeline_layout_1) ordinal(0) : i32
```

Reply:
It's possible so long as they are compatible - this could let us have multiple exported entry points that share some subset of buffers but have differing push constant counts, etc (so two top level functions that fetch push constants/buffers, then a common function that can still reference the base push constants/buffers). We could have a verifier on the executable that maybe poked in and looked for them, but it's likely going to need to be a dedicated pass given how much of the IR tree it pulls together. We'd also then have to codify what pipeline compatibility is and I'm not yet sure we know if we even want to allow that so I'm punting on it for now :P

Reply:
Yeah I figured the analysis would be complex enough that it would need to be a full pass and not a local verifier.